src/goto-programs/remove_function_pointers.cpp

```diff
@@ -521,6 +538,110 @@ void remove_function_pointerst::remove_function_pointer(
 
 /*******************************************************************\
 
+Function: remove_function_pointerst::does_program_contain_const_removal_cast
+
+  Inputs:
+   goto_program - the goto program to check
+
+ Outputs: Returns true if any instruction in the code casts away either
+          explicitly or implicitly the const qualifier of a type.
+
+ Purpose: A naive check to look for casts that remove const-ness from
+          pointers. If this is present, then our remove_const_function_pointerst
+          is not sound so we don't allow it.
+
+\*******************************************************************/
+
+bool remove_function_pointerst::does_program_contain_const_removal_cast(
+  const goto_programt &goto_program)
+{
+  for(const goto_programt::instructiont &instruction :
+    goto_program.instructions)
+  {
+    if(instruction.is_assign())
+    {
+      const code_assignt assign=to_code_assign(instruction.code);
+      typet rhs_type=assign.rhs().type();
+      typet lhs_type=assign.lhs().type();
+
+      // Compare the types recursively for a point where the rhs is more
+      // const that the lhs
+      if(!is_type_at_least_as_const_as(lhs_type, rhs_type))
+      {
+        return true;
+      }
+
+      // see if the rhs looses const inside the expression tree
+      exprt rhs=assign.rhs();
+      while(!rhs.operands().empty())
+      {
+        typet &pre_op_type=rhs.type();
+        typet &post_op_type=rhs.op0().type();
+
+        // Types equality does not check, for example, const-ness
+        // If this is true, then this expression only modified the
+        // type by some qualifier (or not at all)
+        if(pre_op_type==post_op_type)
```

Comment:
True, but note that the code over there does filter out `type1.id()==ID_symbol` (and `type2.id()==ID_symbol`) early on. So what you need to be careful about in comparing types is that some part of them may require lookups in the symbol table. `base_type_eq` takes care of this for you.